
# Project: Investigate TMDb movie data- Part 4

## Table of Contents
<ul>
<li><a href="#intro">I) Introduction</a></li>
<li><a href="#wrangling">II) Data Wrangling</a></li>
<li><a href="#eda">III) Limitations of the Dataset</a></li>
</ul>

# <a id='intro'></a>
## I) Introduction

> This is the part-4 of the report analyzing the TMDB 500 movies dataset. Following questions have been described in this section of the analysis followed by some conclusion and a summary. 



### Questions:





### 9. Outliers : Find the outliers values for runtime, vote_average and budget of the dataset.

           
 

### 10. Limitations of the analysis: Discuss the limitations of the analysis.


### 11. Summary


### 12. Conclusion


### 13. General Reference
               
               
          
               
           
           

### Data properties

In [1]:
# Importing packages 

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

import datetime as dt

%matplotlib inline

## reading the CSV file

df_movies = pd.read_csv('C:\\Users\\raz37388\\Desktop\\Udacity_assignment\\Assignment 2\\tmdb-movies.csv') 


In [2]:
#dimension of the data
df_movies.shape

(10866, 21)

In [3]:
# find duplicate rows
df_movies[df_movies.duplicated()]

id    imdb_id  popularity    budget  revenue original_title  \
2090  42194  tt0411951     0.59643  30000000   967000         TEKKEN   

                                                   cast homepage  \
2090  Jon Foo|Kelly Overton|Cary-Hiroyuki Tagawa|Ian...      NaN   

              director              tagline     ...       \
2090  Dwight H. Little  Survival is no game     ...        

                                               overview runtime  \
2090  In the year of 2039, after World Wars destroy ...      92   

                                           genres    production_companies  \
2090  Crime|Drama|Action|Thriller|Science Fiction  Namco|Light Song Films   

     release_date vote_count  vote_average  release_year  budget_adj  \
2090    3/20/2010        110           5.0          2010  30000000.0   

      revenue_adj  
2090     967000.0  

[1 rows x 21 columns]

In [4]:
#dropping the duplicate 
df_movies.drop_duplicates(inplace=True)

In [5]:
# Checking for the duplicate- should return False 
df_movies.duplicated().any()

False

In [6]:
# View first 5 rows of the dataset
df_movies.head(2)

id    imdb_id  popularity     budget     revenue      original_title  \
0  135397  tt0369610   32.985763  150000000  1513528810      Jurassic World   
1   76341  tt1392190   28.419936  150000000   378436354  Mad Max: Fury Road   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   

                        homepage         director             tagline  \
0  http://www.jurassicworld.com/  Colin Trevorrow   The park is open.   
1    http://www.madmaxmovie.com/    George Miller  What a Lovely Day.   

       ...                                                overview runtime  \
0      ...       Twenty-two years after the events of Jurassic ...     124   
1      ...       An apocalyptic story set in the furthest reach...     120   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...     6/9/2015       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...    5/13/2015       6185   

   vote_average  release_year   budget_adj   revenue_adj  
0           6.5          2015  137999939.3  1.392446e+09  
1           7.1          2015  137999939.3  3.481613e+08  

[2 rows x 21 columns]

In [7]:
## Converting the budget into million USD
df_movies['budget'] = df_movies['budget'] /10**6 

In [8]:
## Budget into million USD
df_movies['budget'].head(3)

0    150.0
1    150.0
2    110.0
Name: budget, dtype: float64

### 9- Outliers 

In [9]:
# Function to find outliers from 25 and 75 percetile values
def find_outliers (df_column):
    
    # input: column of a dataframe
    # outpuer: first_quartile, third_quartile, IQR, lower_limit, upper_limit of the dataframe
    
    first_quartile = np.percentile(df_column, 25)
    
    third_quartile = np.percentile(df_column, 75)
    
    IQR = third_quartile-first_quartile
    
    upper_limit = third_quartile + 1.5*IQR
    
    lower_limit = first_quartile -1.5*IQR
    
    return first_quartile, third_quartile, IQR, lower_limit, upper_limit

### 9-A) Outliers for Runtime

In [10]:
# number of movies with 0 runtime
df_movies[df_movies['runtime']== 0].runtime.count()

31

In [11]:
#nonzero runtime_movies
df_nonzero_runtime_movies = df_movies[df_movies['runtime']>0]

In [12]:
# # number of movies with 0 runtime
df_nonzero_runtime_movies.runtime.count()

10834

In [13]:
#summary statistics for nonzero runtime movies
df_nonzero_runtime_movies.describe()

id    popularity        budget       revenue       runtime  \
count   10834.000000  10834.000000  10834.000000  1.083400e+04  10834.000000   
mean    65750.128854      0.647762     14.665093  3.994086e+07    102.363855   
std     91819.986178      1.001204     30.948988  1.171561e+08     30.948225   
min         5.000000      0.000065      0.000000  0.000000e+00      2.000000   
25%     10586.250000      0.208536      0.000000  0.000000e+00     90.000000   
50%     20551.000000      0.384690      0.000000  0.000000e+00     99.000000   
75%     75055.000000      0.715448     15.000000  2.417932e+07    112.000000   
max    417859.000000     32.985763    425.000000  2.781506e+09    900.000000   

         vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  10834.000000  10834.000000  10834.000000  1.083400e+04  1.083400e+04  
mean     217.962064      5.976343   2001.295274  1.759896e+07  5.151599e+07  
std      576.370933      0.935047     12.819708  3.434421e+07  1.448190e+08  
min       10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%       17.000000      5.400000   1995.000000  0.000000e+00  0.000000e+00  
50%       38.000000      6.000000   2006.000000  0.000000e+00  0.000000e+00  
75%      146.000000      6.600000   2011.000000  2.094246e+07  3.389667e+07  
max     9767.000000      9.200000   2015.000000  4.250000e+08  2.827124e+09

### Note:

From the standard deviation, it can be observed that the movies have a large standard deviation which indicates that the movies are very spread when it comes to runtime.

In [14]:
# Finding the time that would be considered outliers for runtime
runtime_outlier = find_outliers(df_movies['runtime'])

In [15]:
# prints the 0.25 quartile, 0.75 quartiles, interquartile range, lower limit and upper limit
runtime_outlier 

(90.0, 111.0, 21.0, 58.5, 142.5)

In [16]:
# median of the nonzero runtime movies 
df_nonzero_runtime_movies['runtime'].median()

99.0

In [17]:
#Number of outliers-movies  having runtime smaller than 79 and larger than 123 minutes. 
runtime_outliers = df_movies[df_movies['runtime']<= 58.5].original_title.count() + df_movies[df_movies['runtime']>= 142.5].original_title.count()
print(runtime_outliers )

781


### 9-A) Conclusion:

1) 25% movies have runtime 90 minutes or lower

2) 75% movies have runtime 111 minutes or lower

3) Movies with a runtime smaller than 58.5 and larger than 142 minutes will be considered as outliers. 

4) The median values of the runtime are 99 minutes- to 50% of the movies have runtime below 99 minutes and the rest 50% have runtime above 99 minutes.

5) A total of 781 movies fall into the outliers category

### 9-B) Outliers for Vote_average

In [18]:
# number of movies with 0 vote_average- returns 0 if all movies have the vote_average data available
vote_average_outlier = find_outliers(df_movies['vote_average'])

In [19]:
# prints the 0.25 quartile, 0.75 quartiles, interquartile range, lower limit and upper limit
vote_average_outlier

(5.4, 6.6, 1.1999999999999993, 3.6000000000000014, 8.399999999999999)

In [20]:
# median of the nonzero runtime movies 
df_movies['vote_average'].median()

6.0

In [21]:
#Number of outliers-movies  having runtime vote_average smaller than 5.4 and larger than 6.6 
vote_average_outliers = df_movies[df_movies['vote_average']<= 3.6].original_title.count()+ df_movies[df_movies['vote_average']> 8.39].original_title.count() 
print(vote_average_outliers)

197


### 9-B) Conclusion

1) 25% movies have vote_average 5.4  or lower

2) 75% movies have vote_average 6.6  or lower

3) Movies with a vote_average smaller than 3.6 and larger than 8.39 will be considered as outliers. 

4) The median values of the vote_average are 6.0 - to 50% of the movies have vote_average below 6.0  and the rest 50% have vote_average above 6.0

5) A total of 197 movies fall into the category of outliers when it comes to vote_average

### 9-C) Outliers for Budget

In [22]:
# median of the nonzero runtime movies 
df_movies['budget'].median()

0.0

### Note:
Since there is are a large number of movies with 0 budget- that makes the median budget 0 which is unrealisitc.

In [23]:
#taking the nonzero budget_movies
df_nonzero_budget_movies = df_movies[df_movies['budget']>0]

In [24]:
#checking if thethe nonzero budget_movies have any 0 budget movie- should return 0
df_nonzero_budget_movies[df_nonzero_budget_movies['budget'] == 0]

Empty DataFrame
Columns: [id, imdb_id, popularity, budget, revenue, original_title, cast, homepage, director, tagline, keywords, overview, runtime, genres, production_companies, release_date, vote_count, vote_average, release_year, budget_adj, revenue_adj]
Index: []

[0 rows x 21 columns]

In [25]:
# Number of movies with non-zero budget
df_nonzero_budget_movies.budget.count()

5169

In [26]:
# Summary statistics of movies with non-zero budget
df_nonzero_budget_movies.describe().budget

count    5169.000000
mean       30.739576
std        38.904414
min         0.000001
25%         6.000000
50%        17.000000
75%        40.000000
max       425.000000
Name: budget, dtype: float64

In [27]:
# Finding the amount (million USD) that would be considered outliers for budget
budget_outlier = find_outliers(df_nonzero_budget_movies['budget'])

In [28]:
# prints the 0.25 quartile, 0.75 quartiles, interquartile range, lower limit and upper limit
budget_outlier

(6.0, 40.0, 34.0, -45.0, 91.0)

In [29]:
lower_quartiles_number = df_nonzero_budget_movies[df_nonzero_budget_movies['budget']<= 6].budget.count()
upper_quartiles_number = df_nonzero_budget_movies[df_nonzero_budget_movies['budget']>= 91].budget.count()
print ("Number of movies less then 6.0 million USD budget: {}".format(lower_quartiles_number))
print ("Number of movies less then 91.0 million USD budget: {}".format(upper_quartiles_number))

Number of movies less then 6.0 million USD budget: 1355
Number of movies less then 91.0 million USD budget: 352


### 9-C) Conclusion

The lower limit of the outlier value is negative (not possible), which indicates that the difference between the upper bound and the lower bound is too high. Without adjusting the inflation, it is impossible to get a meaningful value. 

2. 1355 movies have a budget lower than the 25% quartile budget of 6 million USD

3. 352 movies have a budget more than the 75% quartile budget of 91 million USD

### 10. Limitations

In [30]:
# Dimension of the dataframe
df_movies.shape

(10865, 21)

In [31]:
# view missing value count for each feature of the dataset
df_movies.isnull().sum()

id                         0
imdb_id                   10
popularity                 0
budget                     0
revenue                    0
original_title             0
cast                      76
homepage                7929
director                  44
tagline                 2824
keywords                1493
overview                   4
runtime                    0
genres                    23
production_companies    1030
release_date               0
vote_count                 0
vote_average               0
release_year               0
budget_adj                 0
revenue_adj                0
dtype: int64

In [32]:
# View sumary of the data
df_movies.describe()

id    popularity        budget       revenue       runtime  \
count   10865.000000  10865.000000  10865.000000  1.086500e+04  10865.000000   
mean    66066.374413      0.646446     14.624286  3.982690e+07    102.071790   
std     92134.091971      1.000231     30.914285  1.170083e+08     31.382701   
min         5.000000      0.000065      0.000000  0.000000e+00      0.000000   
25%     10596.000000      0.207575      0.000000  0.000000e+00     90.000000   
50%     20662.000000      0.383831      0.000000  0.000000e+00     99.000000   
75%     75612.000000      0.713857     15.000000  2.400000e+07    111.000000   
max    417859.000000     32.985763    425.000000  2.781506e+09    900.000000   

         vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  10865.000000  10865.000000  10865.000000  1.086500e+04  1.086500e+04  
mean     217.399632      5.975012   2001.321859  1.754989e+07  5.136900e+07  
std      575.644627      0.935138     12.813260  3.430753e+07  1.446383e+08  
min       10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%       17.000000      5.400000   1995.000000  0.000000e+00  0.000000e+00  
50%       38.000000      6.000000   2006.000000  0.000000e+00  0.000000e+00  
75%      146.000000      6.600000   2011.000000  2.085325e+07  3.370173e+07  
max     9767.000000      9.200000   2015.000000  4.250000e+08  2.827124e+09

In [33]:
# Number of movies with 0 runtime 
df_movies.query('runtime==0').count().runtime

31

In [34]:
# Number of movies with 0 budget
df_movies.query('budget==0').count().budget

5696

In [35]:
# Percentage of missing data in budget
df_movies.query('budget==0').count().budget/df_movies.shape[0]* 100

52.425218591808566

In [36]:
# Number of movies with 0 revenue
df_movies.query('revenue==0').count().revenue

6016

In [37]:
# Percentage of missing data in revenue
df_movies.query('revenue==0').count().revenue/df_movies.shape[0]* 100

55.37045559134837

### 11. Summary

1) The year with the highest number of movies: September 2014, the number of movies released 98

2) The decade with the highest number of movies in the top 500: 2010s (2010-2015)

3) Most frequent cast: Robert De Nero, appeared in 72 movies

4) Most successful cast based on the highest number of movies in the top 500 movies: Tom Hanks, out of 46 movies in the dataset, 9 made it to top 500

5) The director who made the highest number of movies: Woody Allen, produced 46 movies. 

6) Director with the most number of movies in the top 500 movies: 

7) The movie with the longest runtime and highest vote_average: The Story of Film: An Odyssey, a documentary series, has a duration of 900 minutes

8) The movie with the highest budget: The Warrior's Way,  total budget: 425 million USD. 

9) The movie with the highest profit: Avatar by James Cameron

10) Most common genre: Drama

11) The largest number of movies by the production company: Universal Pictures, produced 522 movies

 

### 12. Conclusion

1. In this project, the focus is on exploratory data analysis using some basic descriptive statistics, plots and charts. Hence, no inference can be made based on the correlations in order to come up with criteria to analyze the impact of a specific factor (or combination of factors) for the successful movies with higher profit and higher average rating.


2. Using this basic analysis, no accurate prediction can be made on what runtime, cast, the release date should be combined for a particular genre of movie(s). For that, more rigorous modelling needs to be done with train and test dataset. 


3. In the dataset, there are some missing values in the features that we have used to answer questions on- director, cast, genres, production companies etc.. These missing values can contribute erroneous or biased result in questions like the most frequent cast, most common genres, highest movies by a production company and the highest number of movies by a director. Particularly for the 'production_comapnies' feature, a large number of values are missing. Missing of values could have been dealt with excluding all the rows with null values at the beginning. This, however, would have erased column (features) with nonzero features. Hence, null values have been removed when that particular column was under consideration.There is just one duplicate value in the dataset and that has been removed in all the analysis. 


4. In the dataset, there are a large number of 0 values that make no sense- such as runtime, budget, revenue. Particularly, for the budget and runtime, a large number of values are '0'- 52.42% and 55.36% respectively. This can lead to erroneous result when it comes to finding the movies with the largest runtime, runtime trend per decade, movies with the largest budget and revenue etc. 


5. We can see there is a large number of outliers when it comes to the 'runtime' and 'vote_average' and 'budget' (See section 9-A, 9-B and 9-C). This analysis also includes the outliers when it comes to analyzing the 'runtime', 'vote_average' and 'budget', and that can impact the average rating, average runtime and budget in each decade. 


6. Also, due to the large number of 0 values and not adjusting the inflation the budget outliers cannot be calculated accurately. 




### 13. General Reference
    
    1. Udacity Data Analysis Nanodegree Course Materials
    2. For issues related to python Function consulted:  https://stackoverflow.com
    3. For Python documentation: https://www.python.org/
    